### ISSUE TITLE
Noisy clusters

### ISSUE THREAD

**Issue by shihaisun** [cortex-lab/KiloSort#80]
Hi KiloSort team,

A lot of my clusters are noisy following KiloSort spike-sorting. Below is an example of a noisy cluster (as seen in the feature view and waveform view). 

![image](https://user-images.githubusercontent.com/30403651/29005595-ab2b533c-7b21-11e7-980f-5c3bef61e582.png)

I changed the maximum clusters in the config file to 128 and left the rest on default. Can you guys recommend me any setting changes to fix my noisy clustering issue?

Thank you so much for your time,
Scott

**Comment by nsteinme** [comment]
The big vertical lines on the waveform view suggest that you have some very
large-amplitude artifacts. You might try detecting these with a simple
threshold and then removing that part of the data, then run again. If you
don't think that the artifacts impacted your sorting results much, though,
then you can simply "split" them out in your phy gui. Looks to me like that
would be fine for the example cluster you're showing.

**Comment by brendonw1** [comment]
Hey @nsteinme any nice ways you guys have to run such a filter?  Maybe a plugin?

**Comment by nsteinme** [comment]
No, I don't have a method for this - I always run common average
referencing before sorting, which I guess must be removing any such
artifacts that I have. I rarely see problems like this.

**Comment by brendonw1** [comment]
Ah i see.  That's a good point.  So you create a second binary file that's averaged across channels and then run kilosort on that?  (also, have you ever tried median filtering?).  Thanks

**Comment by nsteinme** [comment]
Yes, I use this function:
https://github.com/cortex-lab/spikes/blob/master/preprocessing/applyCARtoDat.m

No, I haven't tried median filtering...

**Comment by brendonw1** [comment]
Thanks!
By the way it looks like that does 2D median filtering actually! 
 "% Subtracts median of each channel, then subtracts median of each time
% point."
(and all calls are "median" not "mean")

**Comment by nsteinme** [comment]
I think by "median filtering" people sometimes mean like what matlab's "medfilt1" does - where it takes a running window and replaces each point with the median of the values in that window. Here I'm doing something different: First, for each channel separately, I subtract the channel's median (across all time) from every point in the trace (this step just removes channel offsets, which is a thing with neuropixels probes but probably irrelevant for most probes); second, for each time point separately, I take the median across channels and subtract it from every channel's point. The second step is what people call common average referencing (whether it's a mean or median, probably doesn't matter too much...). I should have made that comment in the function more clear!

**Comment by Alkohost** [comment]
Hello!
I just started using Kilosort and faced with noisy clusters also. I mean all detected clusters had no gaps around zero at autocorrelograms. I used [Klusters, NeuroScope, NDManager] - tools before with KlustaKwik for a few channels (usually 4), and there same data had more clear clusters. Can you please tell me what parameters should I change for cluster clearing?

**Comment by g874259** [comment]
Hi @nsteinme,
1. Should we do the data preprocessing (subtract channel's median and CAR) ONLY for good channels?
2. The data preprocessing on 1. should be done after high-pass (300Hz) filtering?
Tks, JJ
